### PR TITLE
Adding VAT number options to settings & checkout

### PR DESCRIPTION
**Pull request by qodo**
# Adding VAT Number options to checkout. #108 

Working with @katalysis to adding some VAT Number options. This adds a new option under **Settings** that allows the user to provide a VAT Number and further options under each Tax Rate whether a tax should not be applied if the user has entered a valid VAT Number.

When a VAT number is applied it is validated against a regular expression first and then secondly it attempts to validate this using the Europa VAT service. If this service fails it falls back to just regular expression validation.

This adds a new attribute to the user for the VAT number and on the checkout process this adds a new field for them to enter a VAT Number.

### CODE REVIEW

File: single_pages/dashboard/store/settings/tax.php [hunk 1]
```diff
@@ -1,105 +1,122 @@
...
+                <div class="col-sm-12">
+
+                    <?php if (Config::get('community_store.vat_number')) { ?>
+                    <div class="form-group">
+                        <div class="checkbox">
+                        <label for="taxVatExclude" class="control-label">
+                            <?= $form->checkbox('taxVatExclude', 1, $taxRate->getTaxVatExclude()); ?>
+                            <strong><?= t("If the EU VAT Number attribute is empty or invalid.")?></strong>
+                        </label>
+                        <p class="help-block">If the customer has entered a valid VAT Number then this tax will not be applied in checkout.</p>
```
Comment (Remo) [inline single_pages/dashboard/store/settings/tax.php]: you should wrap this using the t-function

**Comment by Remo** [comment]
Great to see that someone is working on this!

**Comment by Mesuva** [comment]
This sounds great, a huge amount of work. 
I'll look to merge this in soon.

File: single_pages/dashboard/store/settings/tax.php [hunk 2]
```diff
@@ -1,105 +1,122 @@
...
+                </div>
+                </div>
+            </div>
+
+            <h4><?= t("VAT Number Options")?></h4>
+            <div class="row">
+
+                <div class="col-sm-12">
+
+                    <?php if (Config::get('community_store.vat_number')) { ?>
```
Comment (Mesuva) [inline single_pages/dashboard/store/settings/tax.php]: If the VAT option isn't shown, the VAT Number Options label is still shown - so this check should include line 90

**Comment by Mesuva** [comment]
Just an observation or potential issue from testing this:

When going through the checkout, with the VAT option on, I wasn't able to input blank or something invalid as the VAT number and then progress, even if I had "If the EU VAT Number attribute is empty or invalid." checked against the only tax rate.  I still couldn't if I removed the tax rate and had none defined.

Should I be able to progress through the checkout both with an without a valid VAT number when it's enabled in the global tax settings?

**Comment by Mesuva** [comment]
Just circling back to this - I can see some tweaks have been made code wise, but I still have my original question - I think there still might be a bug (or a misunderstanding) as to how this feature works.

**Comment by qodo** [comment]
Latest changes move the VAT number to AFTER the shipping method and allow the user to leave the field blank. Another issue with "Previous" button was found and fixed. Should have added a separate PR but still getting to grips with Git.

**Comment by Mesuva** [comment]
I just found a moment to test this again.

One thing I'm still finding a bit confusing is the language on this option here:
screen shot 2017-02-20 at 8 49 10 pm

The label on the checkbox sounds like the opposite to the description underneath, so I'm left a bit confused as to what it actually means. Could this be reworded to be clearer as to what it does when checked?

**Comment by Mesuva** [comment]
Sorry I haven't had a chance test/merge this yet, until I do have a few moments (which may be this weekend) can someone else give this some tests?

**Comment by Mesuva** [comment]
I've merged this branch into the master via a testing branch.

I had to adjust a few things for styling, but I was able to test it with and without a valid vat number.

One question I had though - is the entered vat number supposed to show in the order details within the dashboard or in the order notification emails?  Is it supposed to actually be stored against the order or it is purely needed as a test at checkout time?

**Pull request closed without merging**